packages/cli/src/modules/breaking-changes/breaking-changes.service.ts

```diff
@@ -0,0 +1,210 @@
+import { Logger } from '@n8n/backend-common';
+import { WorkflowRepository } from '@n8n/db';
+import { Container, Service } from '@n8n/di';
+import { ErrorReporter } from 'n8n-core/src/errors/error-reporter';
+import { strict as assert } from 'node:assert';
+
+import { RuleRegistry } from './breaking-changes.rule-registry.service';
+import { allRules, RuleInstances } from './rules';
+import type {
+	DetectionReport,
+	BreakingChangeVersion,
+	DetectionResult,
+	IBreakingChangeRule,
+	AffectedWorkflow,
+	IBreakingChangeWorkflowRule,
+} from './types';
+import { BreakingChangeSeverity } from './types';
+import { N8N_VERSION } from '../../constants';
+
+import { CacheService } from '@/services/cache/cache.service';
+
+@Service()
+export class BreakingChangeService {
+	private readonly batchSize = 100;
+	private static readonly CACHE_KEY_PREFIX = 'breaking-changes:results:';
+
+	constructor(
+		private readonly ruleRegistry: RuleRegistry,
+		private readonly workflowRepository: WorkflowRepository,
+		private readonly cacheService: CacheService,
+		private readonly logger: Logger,
+		private readonly errorReporter: ErrorReporter,
+	) {
+		this.logger = logger.scoped('breaking-changes');
+		this.registerRules();
+	}
+
+	registerRules() {
+		const rulesServices: RuleInstances[] = allRules.map((rule) =>
+			Container.get<RuleInstances>(rule),
+		);
+		this.ruleRegistry.registerAll(rulesServices);
+	}
+
+	private async getAllWorkflowRulesResults(
+		workflowLevelRules: IBreakingChangeWorkflowRule[],
+	): Promise<DetectionResult[]> {
+		const totalWorkflows = await this.workflowRepository.count();
+		const allAffectedWorkflows: Map<string, AffectedWorkflow[]> = new Map();
+		const allResults: DetectionResult[] = [];
+
+		this.logger.info('Processing workflows in batches', {
+			totalWorkflows,
+			batchSize: this.batchSize,
+		});
+
+		// Process workflows in batches
+		for (let skip = 0; skip < totalWorkflows; skip += this.batchSize) {
+			const workflows = await this.workflowRepository.find({
+				select: ['id', 'name', 'active', 'nodes'],
+				skip,
+				take: this.batchSize,
+			});
+
+			this.logger.debug('Processing batch', {
+				skip,
+				workflowsInBatch: workflows.length,
+			});
+
+			for (const rule of workflowLevelRules) {
+				try {
+					const ruleId = rule.getMetadata().id;
+					for (const workflow of workflows) {
+						const workflowDetectionResult = await rule.detectWorkflow(workflow);
```

Comment:
This is a lot of looping: rules --> workflows --> rules * nodes. One option could be to group the nodes by their type before passing them along to the rule. That way they could index directly. Other option would be use a traverse pattern and traverse the workflow once and let the rules hook into the traverser